src/app/rpmostree-compose-builtin-rojig.c

```diff
@@ -0,0 +1,615 @@
+  if (!self->corectx)
+    return FALSE;
+
+  const char *arch = dnf_context_get_base_arch (rpmostree_context_get_dnf (self->corectx));
+  self->treefile_rs = ror_treefile_new (treefile_path, arch, self->workdir_dfd, error);
+  if (!self->treefile_rs)
+    return glnx_prefix_error (error, "Failed to load YAML treefile");
+
+  self->treefile_parser = json_parser_new ();
+  if (!json_parser_load_from_data (self->treefile_parser,
+                                   ror_treefile_get_json_string (self->treefile_rs), -1,
+                                   error))
+    return FALSE;
+
+  self->treefile_rootval = json_parser_get_root (self->treefile_parser);
+  if (!JSON_NODE_HOLDS_OBJECT (self->treefile_rootval))
+    return glnx_throw (error, "Treefile root is not an object");
+  self->treefile = json_node_get_object (self->treefile_rootval);
+  self->treespec = rpmostree_composeutil_get_treespec (self->corectx,
+                                                       self->treefile_rs,
+                                                       self->treefile,
+                                                       error);
+
+  *out_context = g_steal_pointer (&self);
+  return TRUE;
+}
+
+static gboolean
+impl_rojig_build (RpmOstreeRojigCompose *self,
+                  const char            *outdir,
+                  gboolean        *out_changed,
+                  GCancellable    *cancellable,
+                  GError         **error)
+{
+  *out_changed = FALSE;
+
+  if (getuid () != 0)
+    {
+      g_printerr ("NOTICE: Running this command as non-root is currently known not to work completely.\n");
+      g_printerr ("NOTICE: Proceeding anyways.\n");
+      g_usleep (3 * G_USEC_PER_SEC);
+    }
+
+  g_autofree char *rojig_name = ror_treefile_get_rojig_name (self->treefile_rs);
+  if (!rojig_name)
+    return glnx_throw (error, "No `rojig` entry in manifest");
+  g_autoptr(GKeyFile) tsk = g_key_file_new ();
+  const char *rojig_output_repo_id = "rpmostree-rojig-output";
+  g_autofree char *rojig_spec = g_strconcat (rojig_output_repo_id, ":", rojig_name, NULL);
+  g_key_file_set_string (tsk, "tree", "rojig", rojig_spec);
+  { const char *repos[] = { rojig_output_repo_id, };
+    g_key_file_set_string_list (tsk, "tree", "repos", (const char*const*)repos, 1);
+  }
+  g_autoptr(RpmOstreeTreespec) treespec = rpmostree_treespec_new_from_keyfile (tsk, error);
+  if (!treespec)
+    return FALSE;
+  g_autoptr(RpmOstreeContext) corectx = rpmostree_context_new_tree (self->cachedir_dfd, self->repo, cancellable, error);
+  if (!corectx)
+    return FALSE;
+  DnfContext *dnfctx = rpmostree_context_get_dnf (corectx);
+  if (!glnx_shutil_mkdir_p_at (self->workdir_dfd, "rojig-repos", 0755, cancellable, error))
+    return FALSE;
+  { g_autofree char *repopath = g_strconcat ("rojig-repos/", rojig_output_repo_id, ".repo", NULL);
+    g_autofree char *repo_contents = g_strdup_printf ("[%s]\n"
```

Comment:
Just thinking out loud here: this is just to find the previous rojig RPM, right? Hmm, I wonder if we could avoid this hack somehow by using `dnf_sack_add_cmdline_package()` instead. Another plus is that one doesn't have to `createrepo_c` the outdir first.

Reply:
We'd have to scrape the dir still to find the latest...and at some point one needs to run `createrepo_c` to serve to clients.